Src/Plugins/Security/SEC101_050.IdentifiableNpmAuthorTokenValidator.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Text;
+
+using Microsoft.CodeAnalysis.Sarif.PatternMatcher.Sdk;
+using Microsoft.RE2.Managed;
+
+using Newtonsoft.Json;
+
+namespace Microsoft.CodeAnalysis.Sarif.PatternMatcher.Plugins.Security
+{
+    public class IdentifiableNpmAuthorTokenValidator : DynamicValidatorBase
+    {
+        [ThreadStatic]
+        private static StringBuilder s_sb;
+
+        protected override IEnumerable<ValidationResult> IsValidStaticHelper(IDictionary<string, FlexMatch> groups)
+        {
+            FlexMatch secret = groups["secret"];
+
+            if (groups.TryGetNonEmptyValue("checksum", out FlexMatch checksum))
+            {
+                string randomPart = secret.Value.String.Substring(4, 30);
+                uint checksumValue = Crc32.Calculate(randomPart);
+                string encodedChecksum = Base62EncodeUint32(checksumValue);
+
+                if (checksum.Value != encodedChecksum)
+                {
+                    return ValidationResult.CreateNoMatch();
+                }
+
+                var validationResult = new ValidationResult
+                {
+                    Fingerprint = new Fingerprint
+                    {
+                        Secret = secret.Value,
+                        Platform = nameof(AssetPlatform.Npm),
+                    },
+                    ValidationState = ValidationState.Unknown,
+                };
+
+                return new[] { validationResult };
+            }
+
+            return ValidationResult.CreateNoMatch();
+        }
+
+        protected override ValidationState IsValidDynamicHelper(ref Fingerprint fingerprint,
+                                                                ref string message,
+                                                                IDictionary<string, string> options,
+                                                                ref ResultLevelKind resultLevelKind)
+        {
+            HttpClient client = CreateOrRetrieveCachedHttpClient();
+
+            return NpmAuthorTokenHelper.ValidateTokens(ref fingerprint,
+                                                       ref message,
+                                                       ref resultLevelKind,
+                                                       client);
+        }
+
+        private static string Base62EncodeUint32(uint value, int minimumLength = 6)
```

Comment:
>Base62EncodeUint32 [](http://example.com/codeflow?start=30&length=18)

should we move this to crc class, make it static and add more tests to complete its coverage? what are ur thoughts?

Reply:
This kind of code should exist in Microsoft.Security.Utilities. That package should have API that accepts an arbitrary alphabet (such as a base62 encoding scheme) and provides encoding and decoding capabilities. 

We will need this API for many classes of secret.

Reply:
Looks like that functionality is still pending. May I proceed with Eddy's suggestion and ensure that a task to incorporate this is added to the backlog?

Reply:
We're now using the CustomAlphabetEncoder class to achieve this.